Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp

```diff
@@ -46,9 +50,34 @@ WebPermissionControllerProxy::~WebPermissionControllerProxy()
     m_process.removeMessageReceiver(Messages::WebPermissionControllerProxy::messageReceiverName());
 }
 
-void WebPermissionControllerProxy::query(const WebCore::ClientOrigin& clientOrigin, const WebCore::PermissionDescriptor& descriptor, WebPageProxyIdentifier identifier, CompletionHandler<void(std::optional<WebCore::PermissionState>, bool shouldCache)>&& completionHandler)
+void WebPermissionControllerProxy::query(const WebCore::ClientOrigin& clientOrigin, const WebCore::PermissionDescriptor& descriptor, std::optional<WebPageProxyIdentifier> identifier, WebCore::PermissionQuerySource source, CompletionHandler<void(std::optional<WebCore::PermissionState>, bool shouldCache)>&& completionHandler)
 {
-    auto webPageProxy = m_process.webPage(identifier);
+    WebPageProxy* webPageProxy = nullptr;
+    if (source == WebCore::PermissionQuerySource::Window || source == WebCore::PermissionQuerySource::DedicatedWorker) {
+        ASSERT(identifier);
+        webPageProxy = m_process.webPage(identifier.value());
+    } else {
+        std::optional<WeakHashSet<WebProcessProxy>> processes;
+        if (source == WebCore::PermissionQuerySource::SharedWorker)
+            processes = m_process.sharedWorkerClientProcesses();
+        if (source == WebCore::PermissionQuerySource::ServiceWorker)
+            processes = m_process.serviceWorkerClientProcesses();
+
+        if (!processes)
+            processes = WeakHashSet<WebProcessProxy>();
+        processes.value().add(m_process);
+
+        for (auto& process : processes.value()) {
+            for (auto* potentialWebPageProxy : process.pages()) {
+                if (WebCore::SecurityOriginData::fromURL(URL { potentialWebPageProxy->currentURL() }) != clientOrigin.topOrigin)
+                    continue;
+                if (webPageProxy && webPageProxy->identifier().toUInt64() >= potentialWebPageProxy->identifier().toUInt64())
```

Comment:
Could be '>'.
It would be good to describe this heuristic in the commit message.

Reply:
Good point--Fixed!